util/app/discovery/discovery.go

```diff
@@ -62,6 +64,25 @@ func Discover(ctx context.Context, appPath, repoPath string, enableGenerateManif
 		if kustomize.IsKustomization(base) && IsManifestGenerationEnabled(v1alpha1.ApplicationSourceTypeKustomize, enableGenerateManifests) {
 			apps[dir] = string(v1alpha1.ApplicationSourceTypeKustomize)
 		}
+
+		if (strings.HasSuffix(base, ".yaml") || strings.HasSuffix(base, ".yml")) && IsManifestGenerationEnabled(v1alpha1.ApplicationSourceTypeDirectory, enableGenerateManifests) {
+			f, err := os.Open(path)
+			if err != nil {
+				log.Errorf("failed to open %s: %v", path, err)
+			}
+			defer f.Close()
+
+			decoder := yaml.NewDecoder(f)
+			obj := make(map[string]interface{})
+			err = decoder.Decode(&obj)
+			if err != nil {
+				log.Errorf("failed to decode %s: %v", path, err)
+			} else {
+				if obj["kind"] == "Service" || obj["kind"] == "Deployment" {
+					apps[dir] = string(v1alpha1.ApplicationSourceTypeDirectory)
+				}
```

Comment:
As for letting helm/kustomize override, something of this sorts could be added as an extra check
`if (obj["kind"] == "Service" || obj["kind"] == "Deployment") && apps[dir]!=helmtype && apps[dir]!=kustomizetype`

I also noticed another issue that it assumes the templates dir inside a helm app to be an app. This could be addressed separately by checking if 
1) Either the file is inside a templates dir 
or 
2) If the directory one level above has a Charts.yaml

Althought I'm not sure if this would make the logic for each file ending in .yaml a bit complex.

Open to other approaches/solutions

Reply:
Also yes a better idea would be to check for presence of `kind` , `apiVersion` and `metaData.name`

Reply:
@jannfis reopening this PR. Would these discussed changes give the expected behaviour

Reply:
> this brings up other issues like configmaps, secrets and other such k8s objects being picked up and their directories being marked as app directories

Yes, and that's basically how it's supposed to be. Every directory that has some YAML is potentially one that can be sourced as an Application. Similar to every directory that has a `kustomization.yaml` in it can be sourced as one, whether it makes sense or not.

> Ideally I think this should discover only service or deployment manifests.

I disagree here. Applications can be anything that has Kubernetes manifests. Often times, an Application will not contain any `Deployment` or `Service` resources.

Reply:
Ah ok, with that idea of app discovery I suppose checking for presence of kind, apiVersion and metadata does make sense. 
Fixed tests and added some more coverage as well.